nuxeo-features/nuxeo-elasticsearch/nuxeo-elasticsearch-core/src/main/java/org/nuxeo/elasticsearch/bulk/IndexRequestComputation.java

```diff
@@ -0,0 +1,162 @@
+        DocumentModelList docs = loadDocuments(session, documentIds);
+        ElasticSearchIndexing esi = Framework.getService(ElasticSearchIndexing.class);
+        for (DocumentModel doc : docs) {
+            try {
+                append(new IndexRequest(indexName, DOC_TYPE, doc.getId()).source(esi.source(doc), XContentType.JSON)
+                                                                         .versionType(VersionType.EXTERNAL)
+                                                                         .version(now));
+            } catch (IOException e) {
+                log.error("Cannot build source for document: " + doc.getId(), e);
+                if (! continueOnFailure) {
+                    abort = true;
+                }
+            }
+        }
+    }
+
+    protected void append(IndexRequest indexRequest) {
+        if (bulkRequest.estimatedSizeInBytes() + indexRequest.source().length() > MAX_RECORD_SIZE) {
+            if (bulkRequest.numberOfActions() > 0) {
+                // Create multiple elastic bulk requests when we exceed the record size
+                bulkRequests.add(bulkRequest);
+                bulkRequest = new BulkRequest();
+            } else {
+                log.warn(String.format("Indexing request for doc: %s, is too large: %d, max record size: %d",
+                        indexRequest.id(), indexRequest.source().length(), MAX_RECORD_SIZE));
+            }
+        }
+        bulkRequest.add(indexRequest);
+    }
+
+    @Override
+    public void endBucket(ComputationContext context, int bucketSize) {
+        if (abort) {
+            context.askForTermination();
+            log.error("Terminate computation due to previous error");
+            return;
+        }
+        bulkRequests.add(bulkRequest);
+        String commandId = getCurrentCommand().getId();
+        int i = 0;
+        for (BulkRequest request : bulkRequests) {
+            // rand robin key
+            String key = bucketKey + "-" + i++;
+            DataBucket dataBucket = new DataBucket(commandId, request.numberOfActions(), toBytes(request));
+            context.produceRecord(OUTPUT_1, Record.of(key, BulkCodecs.getDataBucketCodec().encode(dataBucket)));
+        }
+        bulkRequest = null;
+        bulkRequests.clear();
+    }
+
+    protected String getIndexName(CoreSession session, Map<String, Serializable> properties) {
+        if (properties.containsKey(INDEX_OPTION)) {
+            return (String) properties.get(INDEX_OPTION);
+        }
+        ElasticSearchAdmin esa = Framework.getService(ElasticSearchAdmin.class);
+        return esa.getWriteIndexName(esa.getIndexNameForRepository(session.getRepositoryName()));
+    }
+
+    protected byte[] toBytes(BulkRequest request) {
+        BytesStreamOutput out = new BytesStreamOutput();
+        try {
+            request.writeTo(out);
+            ByteArrayOutputStream out2 = new ByteArrayOutputStream();
```

Comment:
You probably don't need a second byte-like stream. Just do `return BytesReference.toBytes(out.bytes())` or something like that.